cryptoprotocol/src/test/java/org/cryptimeleon/incentivesystem/cryptoprotocol/ProtocolIntegrationTest.java

```diff
@@ -0,0 +1,195 @@
+        // define spend amount
+        var spendAmount2 = new BigInteger("23");
+
+        // user tries to generate spend request
+        try
+        {
+            var spendRequest2 = incSys.generateSpendRequest(updatedToken, pkp.getPk(), spendAmount2, ukp, tid2);
+        }
+        catch(IllegalArgumentException e)
+        {
+            System.out.println("Expected exception when trying to spend " + spendAmount2.toString() +" point using token with " + updatedToken.getPoints().toString() + " points: " + e.getMessage());
+        }
+
+        // ensure user token still contains 20 points
+        Assertions.assertEquals(updatedToken.getPoints().getInteger(), new BigInteger(("20")));
+
+        /**
+         * transaction 4: user spends 18 points
+         */
+
+        System.out.println("Testing successful spend transaction.");
+
+        // generate a fresh ID for the spend transaction
+        var tid3 = usedZn.getUniformlyRandomElement();
+
+        // define spend amount
+        var spendAmount3 = new BigInteger("18");
+
+        // user generates spend request
+        SpendRequest spendRequest3 = null;
+        try
+        {
+            spendRequest3 = incSys.generateSpendRequest(updatedToken, pkp.getPk(), spendAmount3, ukp, tid3);
+        }
+        catch(IllegalArgumentException e)
+        {
+            System.out.println("Unexpected exception when trying to spend " + spendAmount3.toString() +" point using token with " + updatedToken.getPoints().toString() + " points: " + e.getMessage());
+        }
+
+        // serialize and deserialize spend request to ensure that serialization does not break anything
+        var serializedSpendRequest3 = spendRequest3.getRepresentation();
+        FiatShamirProofSystem spendDeductProofSystem = new FiatShamirProofSystem(new SpendDeductZkp(incSys.getPp(), pkp.getPk()));
+        var deserializedSpendRequest3 = new SpendRequest(serializedSpendRequest3, incSys.getPp(), spendDeductProofSystem, spendAmount3, tid3);
+
+        // provider handles spend request and generates spend response and information required for double-spending protection (which is discarded on the fly, since not needed in this test)
+        var spendResponse3 = incSys.generateSpendRequestResponse(deserializedSpendRequest3, pkp, spendAmount3, tid3).getSpendResponse();
+
+        // serialize and deserialize spend request to ensure that serialization does not break anything
+        var serializedSpendResponse3 = spendResponse3.getRepresentation();
+        var deserializedSpendResponse3 = new SpendResponse(serializedSpendResponse3, usedZn, incSys.getPp().getSpsEq());
+
+        // user handles spend response
+        updatedToken = incSys.handleSpendRequestResponse(deserializedSpendResponse3, deserializedSpendRequest3, updatedToken, spendAmount3, pkp.getPk(), ukp);
+
+        // ensure that point count of token is 2 = 20-18
+        Assertions.assertEquals(updatedToken.getPoints().getInteger(), new BigInteger(("2")));
```

Comment:
I would like to verify that this token is correct, for example by calling credit-earn on it?

Reply:
Good idea, done.